File: tests/python/test_integration.py
"""
test_integration.py

integration test: runs the full pipeline with mini data
(fixtures) and verify that the result has the correct structure

These tests do not verify exact numbers (this is a job for 
test_golden.py), but they verify:
  - pipeline finishes without errors
  - all expected sheets are generated
  - sheets have the correct number of columns and rows
  - data types are coherent

They need the mini fixtures in tests/fixtures/.

run:
    pytest tests/python/test_integration.py -v
"""

import os
import sys
import pytest
import openpyxl

# conftest.py already adds the CODE_DIR to the sys.path
from ebdtFunctions import GetExpectancyOfBeingDownstreamTarget


# ==== EXPECTED SHEETS IN OUTPUT EXCEL =====

HOJAS_INPUT  = {"pvalue", "fold"}
HOJAS_OUTPUT = {
    "pvalue.select",
    "fold.select",
    "corrPPwithKinases",
    "ratioSigPPoverSignInhSpeci",
    "ProbOfBeingKinaseSubs",
    "PutativeKinaseSubstrates",
    "tableEdgeSubs",
    "nodes.edges",
}
TODAS_LAS_HOJAS = HOJAS_INPUT | HOJAS_OUTPUT


# ==== FIXTURE: EXECUTED PIPELINE ====

@pytest.fixture(scope="module", params=["MCF7", "HL60", "NTERA2"])
def wb_resultado(request, tmp_path_factory):
    """
    Runs the full pipeline with the mini fixture for each cell line and returns
    the resulting workbook. Parametrized: runs once per cell line (MCF7, HL60, NTERA2).

    Uses scope='module' so each pipeline only executes once for all tests of this file.
    """
    import shutil
    from conftest import FIXTURES_DIR
    cell_line = request.param
    run_dir = str(tmp_path_factory.mktemp(f"{cell_line.lower()}_integration_run"))

    shutil.copy(
        os.path.join(FIXTURES_DIR, f"{cell_line}_mini.xlsm"),
        os.path.join(run_dir, f"{cell_line}.xlsm")
    )
    shutil.copytree(
        os.path.join(FIXTURES_DIR, "requiredData"),
        os.path.join(run_dir, "requiredData")
    )

    old_cwd = os.getcwd()
    try:
        os.chdir(run_dir)
        GetExpectancyOfBeingDownstreamTarget(
            inhibitionThreshold  = 0.1,
            ratioThreshold       = 0.2,
            probabilityThreshold = 0.5,
            cellLinesFiles       = [f"{cell_line}.xlsm"],
        )
    finally:
        os.chdir(old_cwd)

    return openpyxl.load_workbook(
        os.path.join(run_dir, f"{cell_line}.xlsm"),
        read_only=True,
        keep_vba=False
    )


# ==== TESTS: SHEETS ====

def test_se_generan_todas_las_hojas_de_output(wb_resultado):
    #pipeline must generate 8 sheets of output
    sheets = set(wb_resultado.sheetnames)
    for hoja in HOJAS_OUTPUT:
        assert hoja in sheets, f"missing sheet: '{hoja}'"


def test_hojas_de_input_se_conservan(wb_resultado):
    #'fold' and 'pvalue' original sheets must be preserved
    for hoja in HOJAS_INPUT:
        assert hoja in wb_resultado.sheetnames


# ==== TESTS: PutativeKinaseSubstrates ====

def test_PutativeKinaseSubstrates_tiene_header_correcto(wb_resultado):
    #the first row of PutativeKinaseSubstrates must be: kinase, n, substrates.
    ws = wb_resultado["PutativeKinaseSubstrates"]
    header = next(ws.iter_rows(min_row=1, max_row=1, values_only=True))
    assert header[0] == "kinase"
    assert header[1] == "n"
    assert header[2] == "substrates"


def test_PutativeKinaseSubstrates_tiene_una_fila_por_quinasa(wb_resultado):
    
    #It must be at least one row of kinase, but not more than the total of the CSV (163).
    
    ws = wb_resultado["PutativeKinaseSubstrates"]
    data_rows = list(ws.iter_rows(min_row=2, values_only=True))
    data_rows = [r for r in data_rows if r[0] is not None]
    assert len(data_rows) >= 1
    assert len(data_rows) <= 163  


def test_PutativeKinaseSubstrates_columna_n_es_numerica(wb_resultado):
    #Column 'n' (number of substrates) must contains integers
    ws = wb_resultado["PutativeKinaseSubstrates"]
    for row in ws.iter_rows(min_row=2, values_only=True):
        if row[1] is not None:
            assert isinstance(row[1], (int, float)), (
                f"Column 'n' is not numerical: {row[1]!r}"
            )


# ==== TESTS: nodes.edges ====

def test_nodes_edges_tiene_header_correcto(wb_resultado):
    #nodes.edges must start with: edge, weight, subs."""
    ws = wb_resultado["nodes.edges"]
    header = next(ws.iter_rows(min_row=1, max_row=1, values_only=True))
    assert header[0] == "edge"
    assert header[1] == "weight"
    assert header[2] == "subs"


# ==== TESTS: corrPPwithKinases ====

def test_corrPPwithKinases_tiene_valores_numericos(wb_resultado):
    """
    data cells of corrPPwithKinases must be numerical
    (Pearson correlations in [-1, 1]) or None if they were not calculated.
    """
    ws = wb_resultado["corrPPwithKinases"]
    rows = list(ws.iter_rows(min_row=2, values_only=True))
    for row in rows[:5]:  # verify the first 5 rows
        for val in row[1:]:
            if val is not None:
                assert isinstance(val, (int, float)), f"No numerical value: {val!r}"
                assert -1.01 <= val <= 1.01, f"out of range correlation: {val}"


# ==== TESTS: ratioSigPPoverSignInhSpeci ====

def test_ratios_entre_0_y_1(wb_resultado):
    #all ratios must be in [0, 1]
    ws = wb_resultado["ratioSigPPoverSignInhSpeci"]
    for row in ws.iter_rows(min_row=2, values_only=True):
        for val in row[1:]:
            if val is not None:
                assert 0 <= val <= 1, f"out of range ratio [0,1]: {val}"


# ==== TESTS: ProbOfBeingKinaseSubs ====

def test_probabilidades_entre_0_y_1(wb_resultado):
    #all probabilities must be in [0, 1]
    ws = wb_resultado["ProbOfBeingKinaseSubs"]
    for row in ws.iter_rows(min_row=2, values_only=True):
        for val in row[1:]:
            if val is not None:
                assert 0 <= val <= 1, f"out of range probability [0,1]: {val}"

File: tests/R/test_integration.R
# test_integration.R

# Integration tests: run the full pipeline with mini data (fixtures)
# and verify that the result has the correct structure.

# Parametrized over MCF7, HL60 and NTERA2 via a helper function + loop.

# Run:
#   testthat::test_file("tests/R/test_integration.R")

source("helpers.R")

HOJAS_OUTPUT_ESPERADAS <- c(
  "pvalue.select", "fold.select",
  "corrPPwithKinases", "ratioSigPPoverSignInhSpeci",
  "ProbOfBeingKinaseSubs", "PutativeKinaseSubstrates",
  "tableEdgeSubs", "nodes.edges"
)


# ==== HELPER: EJECUTAR TESTS DE INTEGRACION PARA UN WORKBOOK ====

run_mini_tests <- function(cell_line, wb) {

  test_that(paste0(cell_line, ": All expected output sheets are generated"), {
    hojas_actuales <- names(wb)
    for (hoja in HOJAS_OUTPUT_ESPERADAS) {
      expect_true(hoja %in% hojas_actuales,
                  label = paste(cell_line, "-missing sheet:", hoja))
    }
  })

  test_that(paste0(cell_line, ": input sheets are preserved"), {
    expect_true("fold"   %in% names(wb))
    expect_true("pvalue" %in% names(wb))
  })

  test_that(paste0(cell_line, ": PutativeKinaseSubstrates has correct header"), {
    ws <- readWorkbook(wb, sheet = "PutativeKinaseSubstrates", colNames = FALSE)
    expect_equal(as.character(ws[1, 1]), "kinase")
    expect_equal(as.character(ws[1, 2]), "n")
    expect_equal(as.character(ws[1, 3]), "substrates")
  })

  test_that(paste0(cell_line, ": PutativeKinaseSubstrates has at least one row by kinase"), {
    ws <- readWorkbook(wb, sheet = "PutativeKinaseSubstrates", colNames = FALSE)
    filas_datos <- ws[-1, ]
    filas_con_datos <- filas_datos[!is.na(filas_datos[, 1]), ]
    expect_true(nrow(filas_con_datos) >= 1)
    expect_true(nrow(filas_con_datos) <= 163)
  })

  test_that(paste0(cell_line, ": column n of PutativeKinaseSubstrates contains no negative integers"), {
    ws <- readWorkbook(wb, sheet = "PutativeKinaseSubstrates", colNames = FALSE)
    valores_n <- suppressWarnings(as.numeric(as.character(ws[-1, 2])))
    valores_n <- valores_n[!is.na(valores_n)]
    expect_true(all(valores_n >= 0))
    expect_true(all(valores_n == floor(valores_n)))
  })

  test_that(paste0(cell_line, ": nodes.edges has a correct header"), {
    ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
    expect_equal(as.character(ws[1, 1]), "edge")
    expect_equal(as.character(ws[1, 2]), "weight")
    expect_equal(as.character(ws[1, 3]), "subs")
  })

  test_that(paste0(cell_line, ": all correlations are in [-1, 1]"), {
    ws <- readWorkbook(wb, sheet = "corrPPwithKinases", colNames = FALSE)
    for (i in 2:min(6, nrow(ws))) {
      for (j in 2:ncol(ws)) {
        val <- suppressWarnings(as.numeric(as.character(ws[i, j])))
        if (!is.na(val)) {
          expect_true(val >= -1.01 & val <= 1.01,
                      label = paste(cell_line, "-out of range correlation:", val))
        }
      }
    }
  })

  test_that(paste0(cell_line, ":all ratios are between [0, 1]"), {
    ws <- readWorkbook(wb, sheet = "ratioSigPPoverSignInhSpeci", colNames = FALSE)
    for (i in 2:min(6, nrow(ws))) {
      for (j in 2:ncol(ws)) {
        val <- suppressWarnings(as.numeric(as.character(ws[i, j])))
        if (!is.na(val)) {
          expect_true(val >= 0 & val <= 1,
                      label = paste(cell_line, "-out of range ratio:", val))
        }
      }
    }
  })

  test_that(paste0(cell_line, ":all probabilities are between [0, 1]"), {
    ws <- readWorkbook(wb, sheet = "ProbOfBeingKinaseSubs", colNames = FALSE)
    for (i in 2:min(6, nrow(ws))) {
      for (j in 2:ncol(ws)) {
        val <- suppressWarnings(as.numeric(as.character(ws[i, j])))
        if (!is.na(val)) {
          expect_true(val >= 0 & val <= 1,
                      label = paste(cell_line, "-out of range probability:", val))
        }
      }
    }
  })
}


# ==== LOOP: RUN FOR EACH CELL LINE ====

old_wd <- getwd()

for (cell_line in c("MCF7", "HL60", "NTERA2")) {

  run_dir <- make_run_dir(
    xlsm_source = file.path(FIXTURES_DIR, paste0(cell_line, "_mini.xlsm")),
    cell_line   = cell_line
  )

  tryCatch({
    setwd(run_dir)
    suppressWarnings(
      GetExpectancyOfBeingDownstreamTarget(
        inhibitionThreshold  = 0.1,
        ratioThreshold       = 0.2,
        probabilityThreshold = 0.5,
        cellLinesFiles       = c(paste0(cell_line, ".xlsm"))
      )
    )
  }, finally = {
    setwd(old_wd)
  })

  wb <- loadWorkbook(file.path(run_dir, paste0(cell_line, ".xlsm")), keepVBA = FALSE)
  run_mini_tests(cell_line, wb)
}
